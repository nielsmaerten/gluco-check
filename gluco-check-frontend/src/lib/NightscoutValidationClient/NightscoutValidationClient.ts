import { validate } from "class-validator";
import { plainToClass } from "class-transformer";
import { NightscoutValidationClientOptions } from "./types";
import { NightscoutValidationEndpointResponseDto } from "./NightscoutValidationClientDto";
import { NightscoutValidationEndpointRequest } from "../types";

export const NSV_PATH = "/api/v1/validate-nightscout";
export const STATUS_OK = 200;

export class NightscoutValidationClient {
  private endpointUrl: string;
  private controller: AbortController | null;
  private signal: AbortSignal | null;
  private requestInProgress: boolean;

  constructor(options: NightscoutValidationClientOptions) {
    this.endpointUrl = options.endpointUrl;
    this.controller = new AbortController();
    this.signal = this.controller.signal;
    this.requestInProgress = false;
  }

  async fetchValidationStatus(
    nightscoutUrl: string,
    nightscoutToken: string
  ): Promise<NightscoutValidationEndpointResponseDto> {
    if (this.requestInProgress === true) {
      this.cancel();
    }
    try {
      this.requestInProgress = true;
      const requestBody: NightscoutValidationEndpointRequest = {
        url: nightscoutUrl,
        token: nightscoutToken,
      };
      const requestBodyRaw = JSON.stringify(requestBody);
      const headers = new Headers();
      headers.append("Content-Type", "application/json; charset=utf-8");

      let response;
      try {
        response = await fetch(`${this.endpointUrl}${NSV_PATH}`, {
          method: "POST",
          body: requestBodyRaw,
          redirect: "follow",
          headers,
          signal: this.signal,
        });
        if (response.status !== 200) {
          throw new Error("Received non-200 response status");
        }
      } catch (e) {
        throw new Error(`Failed to fetch from nsv endpoint: ${e}`);
      }

      const responseJson = await response?.json();
      try {
        const classifiedResponse = plainToClass(
          NightscoutValidationEndpointResponseDto,
          responseJson
        );

        const errors = await validate(classifiedResponse);
        if (errors.length) {
          throw errors;
        }

        return classifiedResponse;
      } catch (e) {
        throw new Error(`Invalid response from nsv endpoint: ${e}`);
      }
    } catch (e) {
      throw e;
    } finally {
      this.requestInProgress = false;
    }
  }

  async cancel() {
    this.controller?.abort();

    this.controller = null;
    this.signal = null;

    this.controller = new AbortController();
    this.signal = this.controller.signal;
  }

  requestIsInProgress(): boolean {
    return this.requestInProgress;
  }
}
