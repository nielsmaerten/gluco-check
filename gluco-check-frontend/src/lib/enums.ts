export enum BloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol/L",
}

export enum NightscoutBloodGlucoseUnit {
  mgdl = "mg/dl",
  mmol = "mmol",
}

// @TODO: once this repo is properly part of the workspace,
// pull this in from gluco-check-core instead
// These are mapped to translations directly by case name,
// additions or changes here need to be reflected there
export enum DiabetesMetric {
  Everything = "everything",
  BloodSugar = "blood sugar",
  InsulinOnBoard = "insulin on board",
  CarbsOnBoard = "carbs on board",
  SensorAge = "sensor age",
  CannulaAge = "cannula age",
  PumpBattery = "pump battery",
  PumpReservoir = "pump reservoir",
}

export enum AvailableLanguage {
  English = "en",
  Dutch = "nl",
}
