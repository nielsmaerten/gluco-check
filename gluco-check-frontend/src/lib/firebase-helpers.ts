import { BloodGlucoseUnit, DiabetesMetric } from "./enums";
import { GlucoCheckUserDocument } from "./types";

export const FIRESTORE_DEFAULT_SET_OPTIONS: firebase.firestore.SetOptions = {
  mergeFields: [
    "defaultMetrics",
    "glucoseUnit",
    "nightscout.token",
    "nightscout.url",
  ],
};

export const DEFAULT_USER_DOCUMENT: GlucoCheckUserDocument = {
  defaultMetrics: [DiabetesMetric.BloodSugar],
  glucoseUnit: BloodGlucoseUnit.mgdl,
  nightscout: {
    token: "",
    url: "",
  },
};

export const getDocumentPathForUser = (user: firebase.User): string => {
  return `users/${user.email}`;
};
