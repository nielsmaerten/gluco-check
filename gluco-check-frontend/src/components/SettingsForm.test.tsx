import React from "react";
import {
  cleanup,
  render,
  waitFor,
  screen,
  act,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import { clone } from "ramda";
import { BloodGlucoseUnit, DiabetesMetric } from "../lib/enums";
import SettingsForm, { returnHandleOpenTokenDialog } from "./SettingsForm";
import { NightscoutValidationClient } from "../lib/NightscoutValidationClient/NightscoutValidationClient";
import {
  mockNsvResponseDtoInvalidToken,
  mockNsvResponseDtoNonNsUrl,
  mockNsvResponseDtoNsNeedsUpgrade,
  mockNsvResponseDtoValid,
} from "../lib/__mocks__/gluco-check";

jest.mock("react-i18next", () => ({
  useTranslation: () => {
    return {
      t: jest.fn().mockImplementation((i) => {
        return i;
      }),
    };
  },
  Trans: function (props: any) {
    return <span>{props.i18nKey}</span>;
  },
}));

afterEach(() => {
  jest.resetAllMocks();
  cleanup();
});

describe("SettingsForm component", () => {
  const mockNsUrl = "https://example.com/ns";
  const mockNsToken = "token123";
  const mockGlucoseUnits = BloodGlucoseUnit.mgdl;
  const mockDefaultMetrics = [DiabetesMetric.BloodSugar];
  const mockOnSubmit = jest.fn();

  beforeEach(() => {
    mockOnSubmit.mockReset();
  });

  it("renders the component", async () => {
    const { container } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
      />
    );
    const tokenField = await screen.findByTestId("settings-form-field-token");
    const urlField = await screen.findByTestId("settings-form-field-url");
    const glucoseUnitsSelect = await screen.findByTestId(
      "settings-form-field-bg"
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(tokenField).toHaveValue(mockNsToken);
    expect(urlField).toHaveValue(mockNsUrl);
    expect(glucoseUnitsSelect).toHaveValue(mockGlucoseUnits);
  });

  it("renders the component when glucose units field is visible", async () => {
    const { container } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
        shouldShowGlucoseUnitsField={true}
      />
    );
    const tokenField = await screen.findByTestId("settings-form-field-token");
    const urlField = await screen.findByTestId("settings-form-field-url");
    const glucoseUnitsSelect = await screen.findByTestId(
      "settings-form-field-bg"
    );
    expect(glucoseUnitsSelect).toMatchSnapshot();
    expect(glucoseUnitsSelect).toHaveValue(mockGlucoseUnits);
  });

  it("submits the form and saves settings", async () => {
    expect.assertions(12);
    mockOnSubmit.mockImplementationOnce(async (data) => {
      expect(data.nightscoutUrl).toBe(mockNsUrl);
      expect(data.glucoseUnit).toBe(mockGlucoseUnits);
      expect(data.nightscoutToken).toMatchInlineSnapshot(`"token123token"`);
      return Promise.resolve();
    });

    render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
        alertAutohideDuration={200}
      />
    );
    expect(screen.getByTestId("settings-form")).toBeInTheDocument();
    const submitButton = await screen.findByTestId("settings-form-submit");
    const tokenField = await screen.findByTestId("settings-form-field-token");
    const checkboxEverything = await screen.getByLabelText("everything", {
      exact: false,
    });
    const checkbox1 = await screen.getByLabelText("insulin on board", {
      exact: false,
    });
    const checkbox2 = await screen.getByLabelText("blood sugar", {
      exact: false,
    });

    await act(async () => {
      await userEvent.click(checkboxEverything);
    });
    expect(checkbox1).toBeDisabled();
    expect(checkbox1).toBeChecked();
    expect(checkbox2).toBeDisabled();
    expect(checkbox2).toBeChecked();

    await act(async () => {
      await userEvent.click(checkboxEverything);
    });
    expect(checkbox1).not.toBeDisabled();
    expect(checkbox2).not.toBeDisabled();

    await act(async () => {
      await userEvent.click(checkbox1);
      await userEvent.click(checkbox2);

      await userEvent.type(tokenField, "token");
      await userEvent.click(submitButton);
    });
    expect(mockOnSubmit).toHaveBeenCalledTimes(1);

    await waitFor(() => {
      expect(screen.getByTestId("settings-form-submitted")).toBeInTheDocument();
    });

    await waitForElementToBeRemoved(() =>
      screen.getByTestId("settings-form-submitted")
    );
  });

  it("cannot save settings when nighscout url is empty", async () => {
    render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl=""
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
        alertAutohideDuration={200}
      />
    );

    const submitButton = await screen.findByTestId("settings-form-submit");

    expect(submitButton).toBeDisabled();
    await userEvent.click(submitButton);
    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  it("cannot save settings when default metrics are empty", async () => {
    render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={[]}
        onSubmit={mockOnSubmit}
        alertAutohideDuration={200}
      />
    );

    const submitButton = await screen.findByTestId("settings-form-submit");

    expect(submitButton).toBeDisabled();
    await userEvent.click(submitButton);
    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  it("attempts to save settings and handles submission error", async () => {
    expect.assertions(3);
    mockOnSubmit.mockRejectedValueOnce(new Error("To err is human"));

    render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
      />
    );
    expect(screen.getByTestId("settings-form")).toBeInTheDocument();
    const submitButton = await screen.findByTestId("settings-form-submit");
    const tokenField = await screen.findByTestId("settings-form-field-token");

    await act(async () => {
      await userEvent.type(tokenField, "token");
      await userEvent.click(submitButton);
    });
    expect(mockOnSubmit).toHaveBeenCalledTimes(1);

    await waitFor(() => {
      expect(screen.getByTestId("settings-form-error")).toBeInTheDocument();
    });
  });

  describe("Form validation", () => {
    const mockNsvUrl = "https://example.com";
    const nsvClient = new NightscoutValidationClient({
      endpointUrl: mockNsvUrl,
    });
    const mockValidationMethodSpy = jest.spyOn(
      nsvClient,
      "fetchValidationStatus"
    );

    beforeEach(() => {
      mockValidationMethodSpy.mockReset();
    });

    it("Can still render when validation request throws", async () => {
      expect.assertions(1);
      mockValidationMethodSpy.mockRejectedValue(new Error());

      await act(async () => {
        const { container } = render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );
        expect(container).toBeDefined();
      });
    });

    it("Can still render when validation request returns nothing", async () => {
      expect.assertions(1);
      // @ts-expect-error
      mockValidationMethodSpy.mockResolvedValueOnce(undefined);

      await act(async () => {
        const { container } = render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );
        expect(container).toBeDefined();
      });
    });

    describe("errors", () => {
      it("Displays error message for empty nightscout url", async () => {
        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl=""
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.nightscoutUrl.required"
          )
        ).toBeTruthy();
      });

      it("Displays error message for empty metrics", async () => {
        mockValidationMethodSpy.mockResolvedValue(mockNsvResponseDtoValid);

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={[]}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.defaultMetrics.required"
          )
        ).toBeTruthy();
      });
    });

    describe("warnings", () => {
      it("Displays warning message for not-a-nightscout site", async () => {
        mockValidationMethodSpy.mockResolvedValue(mockNsvResponseDtoNonNsUrl);

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.nightscoutUrl.notNightscout"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for a nightscout site below minimum version threshold", async () => {
        mockValidationMethodSpy.mockResolvedValue(
          mockNsvResponseDtoNsNeedsUpgrade
        );

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.nightscoutUrl.needsUpgrade"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for empty token", async () => {
        mockValidationMethodSpy.mockResolvedValue(
          mockNsvResponseDtoInvalidToken
        );

        render(
          <SettingsForm
            nightscoutToken={""}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.nightscoutToken.empty"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for invalid token", async () => {
        mockValidationMethodSpy.mockResolvedValue(
          mockNsvResponseDtoInvalidToken
        );

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={mockDefaultMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.nightscoutToken.invalid"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for unsupported metrics, when token is invalid", async () => {
        const allMetrics = [
          DiabetesMetric.BloodSugar,
          DiabetesMetric.CannulaAge,
          DiabetesMetric.CarbsOnBoard,
          DiabetesMetric.InsulinOnBoard,
          DiabetesMetric.PumpBattery,
          DiabetesMetric.SensorAge,
        ];
        mockValidationMethodSpy.mockResolvedValue(
          mockNsvResponseDtoInvalidToken
        );

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={allMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.defaultMetrics.notAvailableInvalidToken"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for unsupported metrics, when token is empty", async () => {
        const allMetrics = [
          DiabetesMetric.BloodSugar,
          DiabetesMetric.CannulaAge,
          DiabetesMetric.CarbsOnBoard,
          DiabetesMetric.InsulinOnBoard,
          DiabetesMetric.PumpBattery,
          DiabetesMetric.SensorAge,
        ];
        mockValidationMethodSpy.mockResolvedValue(
          mockNsvResponseDtoInvalidToken
        );

        render(
          <SettingsForm
            nightscoutToken=""
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={allMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.defaultMetrics.notAvailableInvalidToken"
          )
        ).toBeTruthy();
      });

      it("Displays warning message for unsupported metrics, when token is valid but some metrics are not found", async () => {
        const allMetrics = [
          DiabetesMetric.BloodSugar,
          DiabetesMetric.CannulaAge,
          DiabetesMetric.CarbsOnBoard,
          DiabetesMetric.InsulinOnBoard,
          DiabetesMetric.PumpBattery,
          DiabetesMetric.SensorAge,
        ];
        const alteredResponseDto = clone(mockNsvResponseDtoValid);
        alteredResponseDto.discoveredMetrics = [DiabetesMetric.BloodSugar];
        mockValidationMethodSpy.mockResolvedValue(alteredResponseDto);

        render(
          <SettingsForm
            nightscoutToken={mockNsToken}
            nightscoutUrl={mockNsUrl}
            glucoseUnit={mockGlucoseUnits}
            defaultMetrics={allMetrics}
            onSubmit={mockOnSubmit}
            nightscoutValidator={nsvClient}
            validationDebounceDuration={0}
          />
        );

        expect(
          await screen.findByText(
            "settings.form.helperText.defaultMetrics.notAvailableValidToken"
          )
        ).toBeTruthy();
      });
    });
  });

  it("has no axe violations", async () => {
    const { container } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        defaultMetrics={mockDefaultMetrics}
        onSubmit={mockOnSubmit}
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});

describe("returnHandleOpenTokenDialog", () => {
  it("returns the curried handler", () => {
    const mockHandler = jest.fn();
    const mockState = true;
    const handler = returnHandleOpenTokenDialog(mockState, mockHandler);
    expect(handler).toMatchInlineSnapshot(`[Function]`);
    handler();
    expect(mockHandler).toHaveBeenCalledWith(!mockState);
  });
});
