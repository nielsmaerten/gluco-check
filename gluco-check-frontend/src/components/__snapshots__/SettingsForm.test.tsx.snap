// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm component renders the component 1`] = `
<form
  class="makeStyles-form-2"
  data-testid="settings-form"
>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
      for="settings-form-field-url"
    >
      settings.form.labels.nightscoutUrl
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        data-testid="settings-form-field-url"
        id="settings-form-field-url"
        name="nightscoutUrl"
        type="text"
        value="https://example.com/ns"
      />
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
      for="settings-form-field-token"
    >
      settings.form.labels.nightscoutToken
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
    >
      <div
        class="MuiInputAdornment-root MuiInputAdornment-positionStart"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zm-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2zm3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1 1.71 0 3.1 1.39 3.1 3.1v2z"
          />
        </svg>
      </div>
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
        data-testid="settings-form-field-token"
        id="settings-form-field-token"
        name="nightscoutToken"
        type="text"
        value="token123"
      />
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-filled"
    >
      <button
        class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiLink-button MuiTypography-colorPrimary"
        type="button"
      >
        settings.form.helperText.nightscoutToken.default
      </button>
    </p>
  </div>
  <fieldset
    class="MuiFormControl-root"
    data-testid="settings-form-fieldset-metrics"
  >
    <legend
      class="MuiFormLabel-root"
    >
      settings.form.labels.defaultMetrics
    </legend>
    <div
      class="MuiFormGroup-root makeStyles-checkboxArray-3 MuiFormGroup-row"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.Everything
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-checked-7 Mui-checked MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.BloodSugar
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.InsulinOnBoard
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.CarbsOnBoard
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.SensorAge
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.CannulaAge
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.PumpBattery
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-9"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          diabetesMetrics.PumpReservoir
        </span>
      </label>
    </div>
  </fieldset>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        data-testid="settings-form-field-bg"
        name="glucoseUnit"
        type="hidden"
        value="mg/dl"
      />
    </div>
  </div>
  <div
    class="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
    >
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
          data-testid="settings-form-submit"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            settings.form.submitButton
          </span>
        </button>
      </div>
    </div>
  </div>
</form>
`;

exports[`SettingsForm component renders the component when glucose units field is visible 1`] = `
<input
  aria-hidden="true"
  class="MuiSelect-nativeInput"
  data-testid="settings-form-field-bg"
  name="glucoseUnit"
  tabindex="-1"
  value="mg/dl"
/>
`;
